WordPress/wordpress-develop
# Add required logic for Block template themes

Trac ticket: https://core.trac.wordpress.org/ticket/53176

## List of files in Gutenberg, and where to map them to in Core

Attempting to fit into Core's existing file hierarchy and naming scheme.

- [x] `lib/template-canvas.php` -> `src/wp-includes/`

`lib/full-site-editing/`:

- [x] `block-templates.php` -> `src/wp-includes/block-template-loader.php` (to avoid confusion with `block-template.php`)
  - Loads block templates from HTML block template file or `wp_template` object.
- [x] `class-wp-block-template.php` -> `src/wp-includes/`
  -  Class representing a block template. Used by the block template loader, and the REST API controller.
- [x] `class-wp-rest-templates-controller.php` -> `src/wp-includes/rest-api/endpoints/`
  - The REST API controller for templates.
- [x] `default-template-types.php` -> `src/wp-includes/`
  - Titles and descriptions of template types. Arguably not block template specific, but OTOH not used outside of FSE.
- [x] ~`edit-site-export.php` - REST endpoint for exporting the contents of the Site Editor.~ _Not needed for WP 5.8._
- [x] ~`edit-site-page.php` -> maybe `src/wp-admin/`? The Post Editor entry point is there, in `src/wp-admin/edit-form-blocks.php`~ _Not needed for WP 5.8._
  - Site Editor entrypoint.
- [x] `full-site-editing.php`
  - [x] Some helper functions to determine is the current theme supports block templates -> `src/wp-includes/theme.php`
  - [ ] ~plus some wp-admin customizations to remove legacy theme screens and sidebar menu items _if the Site Editor is enabled_~ _Not needed for WP 5.8._
- [x] `page-templates.php` -> `src/wp-includes/class-wp-theme.php`
  - Allow to select a block template as custom page template for a given page. See https://github.com/WordPress/gutenberg/pull/27778 for details.
  - Uses `gutenberg_get_block_templates`, which is defined in the block template loader.
  - Hooks into the `theme_templates` filter, which is applied in Core's `src/wp-includes/class-wp-theme.php`. Move there?
- [x] `template-loader.php` -> `src/wp-includes/block-template.php`
  - Modify the template hierarchy resolution algorithm (by hooking into the `{type}_template` filter) to include block templates.
  - Should probably be integrated into `get_query_template()` for now.
- [x] ~`template-parts.php` -> `src/wp-includes/post.php`? (It seems like pretty much all of Core's post types are registered there, including e.g. `wp_block`.)~ _Not needed for WP 5.8._
  - Register the `wp_template_part` CPT.
- [x] `templates-utils.php` -> ~`src/wp-admin/` or~ `src/wp-admin/includes/`?
  - Utilities to customize the `wp_template` admin list. Hooked into actions and filters in `templates.php` and `template-parts.php` for those two post types.
- [x] `templates.php`
  - [x] Register the `wp_template` CPT -> `src/wp-includes/post.php` (It seems like pretty much all of Core's post types are registered there, including e.g. `wp_block`.)
  - [x] Register the `wp_theme` taxonomy -> `src/wp-includes/taxonomy.php`
  - [x] Grant capabilities for the `wp_template` post type to the user -> `src/wp-includes/post.php`
  - [x] Add the Templates wp-admin screen and menu entry (under Themes) -> `src/wp-includes/theme-templates.php`
  - [x] Set unique slug upon auto-draft creation
  - [x] Add the skip link to `wp_footer` -> `src/wp-includes/default_filters.php`

`phpunit/`:

- [x] `class-block-templates-test.php` -> `tests/phpunit/tests/block-template-loader.php`
  -  Unit test coverage for the block-template loading functions (load from file or `wp_template`, respectively).
- [x] ~`class-edit-site-export-test.php`~ _Not needed for WP 5.8._
  - Test coverage for `edit-site-export.php`.
- [x] `class-template-loader-test.php` -> `tests/phpunit/tests/block-template.php`
  - Unit test coverage for the block-template relevant modifications for the template resolution algorithm.
- [x] `class-wp-rest-template-controller-test.php` -> `tests/phpunit/tests/rest-api/rest-templates-controller.php`

`phpunit/fixtures/themes/`:

- [x] `test-theme/` -> `tests/phpunit/data/themedir1/block-template-theme/`
- [x] `test-theme-child/` -> `tests/phpunit/data/themedir1/block-template-theme-child/`

## TODO

- [x] Fix theme loading in `tests/phpunit/tests/block-template.php`.
- [x] Use different theme in `tests/phpunit/tests/rest-api/rest-templates-controller.php` (i.e. not TT1 Blocks, since it's not included with Core) to make it pass. Should ideally be a test block template theme fixture that we're including with `wordpress-develop`. Maybe we can use `tests/phpunit/data/themedir1/block-template-theme/` (or extend it a bit).
- [x] Remove template parts specific bits from the templates REST API controller and its unit test (unless we decide to include template parts in WP 5.8 after all).
- [x] Remove `_gutenberg_` and `gutenberg_` prefixes.
- [x] Make sure the loading order in `src/wp-settings.php` is fine.
- [x] Revisit if we need `if ( gutenberg_supports_block_templates() )` conditionals (e.g. for post type registration in in `post.php`)
- [x] Add/update PHPDocs to cover added/changed functionality; add `@since` items where needed.

## Testing instructions

To run unit tests locally:
```
npm run test:php tests/phpunit/tests/block-template.php
npm run test:php tests/phpunit/tests/block-template-loader.php
npm run test:php tests/phpunit/tests/rest-api/rest-templates-controller.php
```

(or just omit the file args altogether to run all unit tests)

Furthermore, try in your browser: Activate a block template theme (e.g. TT1 Blocks), and verify that it renders properly on the frontend. (Make sure that the GB plugin isn't active!)

Any necessary customizations to GB can be done after the fact and shouldn't be a blocker to this PR.

There's a chance that this list isn't exhaustive -- I might've missed some required files.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**
